cmd/catchupsrv/download.go

```diff
@@ -44,34 +48,83 @@ var connsFlag = flag.Int("conns", 2, "Number of connections per server")
 var serverList []string
 var nextBlk uint64
 
-func blockToString(blk uint64) string {
-	return strconv.FormatUint(blk, 36)
+// padLeftZeros pad the string s with zeros on the left to the length n
+func padLeftZeros(s string, n int) string {
+	return strings.Repeat("0", n-len(s)) + s
+}
+
+// blockToFileName converts a block number into the filename in which it will be downloaded
+// When using subfolders, the string corresponding to the base-36 number padded with zeros
+// Otherwise it is just the base-36 number
+func blockToFileName(blk uint64) string {
+	if *subfoldersFlag {
+		return padLeftZeros(strconv.FormatUint(blk, 36), lenBlockStr)
+	} else {
+		return strconv.FormatUint(blk, 36)
+	}
+}
+
+// stringToBlock converts a base-36 string into a block number
+func stringToBlock(s string) uint64 {
+	blk, err := strconv.ParseUint(s, 36, 64)
+	if err != nil {
+		panic(err)
+	}
+	return blk
+}
+
+// blockToPath converts a block number into the full path in which it will be downloaded
+// When using subfolders, the path is `0b/cd/0bcdef` for block `bcdef`
+// Otherwise the path is just `bcdef` for block `bcdef`
+func blockToPath(blk uint64) string {
+	if *subfoldersFlag {
+		s := blockToFileName(blk)
+		return path.Join(
+			s[0:2],
+			s[2:4],
+			s,
+		)
+	} else {
+		return blockToFileName(blk)
+	}
 }
 
+// stringBlockToPath is the same as blockToPath except it takes a (non-padded) base-36 block
+func stringBlockToPath(s string) string {
+	if *subfoldersFlag {
+		return blockToPath(stringToBlock(s))
+	} else {
+		return s
+	}
+}
+
+// blockDir returns the root folder where all the blocks are stored
 func blockDir() string {
 	return filepath.Join(*dirFlag, fmt.Sprintf("v1/%s/block", *genesisFlag))
 }
 
-func blockFile(blk uint64) string {
-	return filepath.Join(blockDir(), blockToString(blk))
+// blockFullPath returns the full path to a block, including blockDir
+func blockFullPath(blk uint64) string {
+	return filepath.Join(blockDir(), blockToPath(blk))
 }
 
 func blockURL(server string, blk uint64) string {
-	return fmt.Sprintf("http://%s/v1/%s/block/%s", server, *genesisFlag, blockToString(blk))
+	return fmt.Sprintf("http://%s/v1/%s/block/%s", server, *genesisFlag, blockToFileName(blk))
```

Comment:
Is there a reason to left-pad the base-36 block number in the URL?

Reply:
Indeed, this was a bug. Fixed. Thanks!

Reply:
not related to your change, but we should generally avoid formatting URLs. instead, we should construct them by using url.URL, which (should) do the magic of string concatenations.

Reply:
Sorry I'm not completely sure how to use `url.URL` to do that. When I look at the documentation, I do not see a way to format the `url.URL.Path` from multiple components like `path.Join` would do for files. But I may have missed something.